### ISSUE TITLE
No externs are generated for re-exported namespace

### ISSUE THREAD

**Issue by theseanl** [angular/tsickle#1203]
As I understand tsickle does not support `export * from '...'` syntax (which I couldn't find a tracking issue), but it seems that it does not support the following situation as well:
```
import * as foo from 'bar';
export {foo}
```
When this is compiled with tsickle, the externs output is empty.

**Comment by evmar** [comment]
I did some work in this area recently within Google.  My understanding* is that Closure compiler cannot optimize the result of code like this, in that it retains every field on foo whether it's used or not.  So we banned this pattern in our code in general.  I believe if you write code like this you get a Closure compiler warning about it backing off off optimization.

* Maybe here I should disclaim: I know a lot about tsickle and I surely know more about Closure compiler than most, but I know very little about Closure compiler.

**Comment by evmar** [comment]
(You mentioned externs in the bug title but not in the bug body, so I wasn't clear whether you were talking about d.ts here or not.)

**Comment by theseanl** [comment]
Yes, it was about `.d.ts` files. For user-controlled code it is much less of a problem, but not for `.d.ts` file which often comes from third-party. I've edited the issue body to clarify it.

**Comment by evmar** [comment]
In general, I believe tsickle doesn't support imports or exports within externs at all.  I think there's a more general bug in the tracker that is about this problem, that this might be considered a dup or a subtask of.  We don't really have a good conception of how to even represent this, given that externs files themselves can't have import/exports.

**Comment by theseanl** [comment]
> We don't really have a good conception of how to even represent this,

Isn't a module `.d.ts` just externs declared on some namespace? It is the idea I've learned from tsickle.
> I believe tsickle doesn't support imports or exports within externs at all

Module `.d.ts` already work 90% of time. Given that almost all declaration files in DefinitelyTyped are module `.d.ts`, IMO making it more complete would open a lot of use cases.

**Comment by evmar** [comment]
We definitely would like this to work, we just haven't had time for it.  I think you're right that hiding things on a namespace works for a single file, but I'm not sure whether all the various other interactions do the right thing with it.

**Comment by evmar** [comment]
(Part of the reason we haven't worked on this much is that we don't have a good model for using third-party dependencies in JSCompiler even when there is no TS involved.)

**Comment by theseanl** [comment]
> I'm not sure whether all the various other interactions do the right thing with it.

Huh, I thought it is already doing the right thing most of time. If you have some particular concerns I'd like to know, I am willing to work on this issue and provide a PR.

**Comment by evmar** [comment]
Sure, send a PR for this if you have a fix.
